cadc-data-ops-fits/src/main/java/org/opencadc/fits/slice/NDimensionalSlicer.java

```diff
@@ -0,0 +1,399 @@
+     *
+     * @param fitsFile     The File to read bytes from.  This method will not close this file.
+     * @param slices       The string value of the pixels to extract.
+     * @param outputStream Where to write bytes to.  This method will not close this stream.
+     * @throws FitsException    Any FITS related errors from the NOM TAM Fits library.
+     * @throws IOException      Reading Writing errors.
+     */
+    public void slice(final File fitsFile, final Slices slices, final OutputStream outputStream)
+            throws FitsException, IOException {
+        slice(new RandomAccessFileExt(fitsFile, "r"), slices, outputStream);
+    }
+
+    /**
+     * Perform a slice operation from the input RandomAccess.  Implementors will walk through the RandomAccess, skipping
+     * unwanted bytes.
+     *
+     * This implementation relies on the NOM TAM FITS API to be able to look up an HDU by its Extension Name
+     * (<code>EXTNAME</code>) and, optionally, it's Extension Version (<code>EXTVER</code>).  This may affect an
+     * underlying <code>Fits(InputStream)</code> unless the <code>InputStream</code> can handle resetting and marking.
+     *
+     * It is NOT the responsibility of this method to manage the given <code>OutputStream</code>.  The caller will
+     * need to close it and ensure it's open outside the bounds of this method.
+     *
+     * @param randomAccessDataObject         The RandomAccess object to read bytes from.  This method will not close
+     *                                       this file.
+     * @param slices       The string value of the pixels to extract.
+     * @param outputStream Where to write bytes to.  This method will not close this stream.
+     * @throws FitsException    Any FITS related errors from the NOM TAM Fits library.
+     * @throws IOException      Reading Writing errors.
+     */
+    public void slice(final RandomAccessDataObject randomAccessDataObject, final Slices slices,
+                      final OutputStream outputStream)
+            throws FitsException, IOException {
+        final ArrayDataOutput output = new BufferedDataOutputStream(outputStream);
+        final Slices.ExtensionSliceValue[] extensionSliceValues = slices.getExtensionSliceValues();
+
+        LOGGER.debug("Parsed extension slice values: " + Arrays.toString(extensionSliceValues));
+        final Fits fits;
+
+        try {
+            fits = new Fits(randomAccessDataObject);
+        } catch (FitsException fitsException) {
+            throw new IllegalStateException(fitsException.getMessage(), fitsException);
+        }
+
+        slice(fits, extensionSliceValues, output);
+    }
+
+    private void slice(final Fits fits, final Slices.ExtensionSliceValue[] extensionSliceValues,
+                       final ArrayDataOutput output) throws FitsException, IOException {
+        // Assume MEF if more than one extension was requested, so write a Primary HDU before anything else.
```

Comment:
This logic is not going to hold up and one cannot rely on the input requested slices to decide what to output.

I think of FITS files as having 3 broad types:
1. simple FITS file with data in the primary HDU
2. me FITS file with no data in the primary and 1+ extensions
3. me FITS file with data in the primary HDU and 1+ extensions
While 2 and 3 look the same structurally, there is one important distinction: in 2, extensions "inherent" metadata from the primary header -- in 3, no metadata is inherited by extensions and each is self contained... you could probably think about 3 as being "just extensions and no common metadata" and 1 could be thought a special case of 3.

So, the first thing is to determine if this is case 2 or not and that means reading the primary header to find out.

Reply:
Does this suggest that we should write out the same style as the input file?

I can think of two use cases where that would not be appropriate:

1. [0][100:300][0][3000:4000] - Two cutouts from a simple FITS file.  Couldn't this be a possibility with a big cube like VLASS?  This sounds like an MEF.
1. [121][300:400] - A single cutout from an MEF.  This sounds like a simple FITS.

Finally, I'm sure there are cases where there is no overlap from any requested cutout, in which case you would get an MEF (if cutting from one) with just a primary header or (possibly) _just_ a header for each requested cutout.

Reply:
For the first case: yes, that should be an MEF result (1 ext per cutout). You may have to change SIMPLE and add NEXTEND keywords if the input was simple. In the case where you really mean extension 0 (so either type 1 or so type 3 above) the result is also type 3... but in general two cutouts of an extension create two extensions no matter where they happen.

For the second, if the input is type 2 (no data in primary) then metadata in the primary applies (inherits) to the extension and you have to create an MEF with the primary + one extension. If it is of type 3, you ignore do not output [0] and then the result does turn into a simple (so again: care with SIMPLE and NEXTEND if that happens in [0]).

Reply:
**NDimensionalSlicer.getOverlap():**
This *must* iterator through the fits extensions in order and pick out the ones that overlap **in the order** they occur in the file. By using getHDU(fits, slice) you are jumping around in the fits file in the arbitrary request order. The order in the returned array will still end up being request order rather than file order. 

So the loop here must be calling `fits.readHDU()` until it finds all the requested extensions. It will have to read to the end if a requested slice is not in the file, which is a worst case. It is also important for this method to have a comment somewhere that states that the returned array has the extension slices in the correct -- file -- order.

Secondarily, it relies heavily on true random access with backwards seeking so could never be made to work (in a sub-optimal fashion) with a seekable stream. That's not an immediate use case but no sense in doing things that make that worse. Specifically, this code can be done in a way that it could work on a seekable stream with 2 passes through the stream. By allowing backwards seeks, that will be N passes instead of 2.. and they are not necessary and provide no benefit.

Reply:
**NDimensionalSlicer line 250:** wasn't the HDU already found by getOverlap()?

Reply:
**NDimensionalSlicer line 489 (hasData):**

Trying to trace what happens when this is called... `data` should be an ImageData and getSize() calls getTrueSize() which returns  the `byteSize` member. That gets assigned in one of the ctors and I assume in this usage it is the ImageData(Header) ctor, which means parseHeader() computes the byteSize from BITPIX, NAXIS, and NAXISn values. Can you verify that this is what happens?